pkg/adaptor/cloud/ibmcloud-powervs/provider.go

```diff
@@ -0,0 +1,228 @@
+// Copyright Confidential Containers Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package ibmcloud_powervs
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"log"
+	"net"
+	"time"
+
+	"github.com/IBM-Cloud/power-go-client/power/models"
+	"github.com/IBM/go-sdk-core/v5/core"
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/cloud"
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/util"
+	"github.com/confidential-containers/cloud-api-adaptor/pkg/util/cloudinit"
+)
+
+const (
+	maxRetries          = 25
+	pollingInterval     = 5
+	dhcpPollingInterval = 30
+	maxInstanceNameLen  = 63
+)
+
+var logger = log.New(log.Writer(), "[adaptor/cloud/ibmcloud-powervs] ", log.LstdFlags|log.Lmsgprefix)
+
+type ibmcloudPowerVSProvider struct {
+	powervsService
+	serviceConfig *Config
+}
+
+func NewProvider(config *Config) (cloud.Provider, error) {
+
+	logger.Printf("ibmcloud-powervs config: %#v", config.Redact())
+
+	powervs, err := newPowervsClient(config.ApiKey, config.ServiceInstanceID, config.Zone)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ibmcloudPowerVSProvider{
+		powervsService: *powervs,
+		serviceConfig:  config,
+	}, nil
+}
+
+func (p *ibmcloudPowerVSProvider) CreateInstance(ctx context.Context, podName, sandboxID string, cloudConfig cloudinit.CloudConfigGenerator) (*cloud.Instance, error) {
+
+	instanceName := util.GenerateInstanceName(podName, sandboxID, maxInstanceNameLen)
+
+	userData, err := cloudConfig.Generate()
+	if err != nil {
+		return nil, err
+	}
+
+	body := &models.PVMInstanceCreate{
+		ServerName:  &instanceName,
+		ImageID:     &p.serviceConfig.ImageID,
+		KeyPairName: p.serviceConfig.SSHKey,
+		Networks: []*models.PVMInstanceAddNetwork{
+			{
+				NetworkID: &p.serviceConfig.NetworkID,
+			}},
+		Memory:     core.Float64Ptr(p.serviceConfig.Memory),
+		Processors: core.Float64Ptr(p.serviceConfig.Processors),
+		ProcType:   core.StringPtr(p.serviceConfig.ProcessorType),
+		SysType:    p.serviceConfig.SystemType,
+		UserData:   base64.StdEncoding.EncodeToString([]byte(userData)),
+	}
+
+	logger.Printf("CreateInstance: name: %q", instanceName)
+
+	pvsInstances, err := p.powervsService.instanceClient.Create(body)
+	if err != nil {
+		logger.Printf("failed to create an instance : %v", err)
+		return nil, err
+	}
+
+	if len(*pvsInstances) <= 0 {
+		return nil, fmt.Errorf("there are no instances created")
+	}
+
+	ins := (*pvsInstances)[0]
+	instanceID := *ins.PvmInstanceID
+
+	num := 1
```

Comment:
Instead of the for loop can you use retry.Do here with a timeout?

Reply:
Hi @bpradipt, I tried using `retry.Do` as below
```
        ctx, cancel := context.WithTimeout(ctx, 150*time.Second)
	defer cancel()

	logger.Printf("Waiting for instance to reach state: ACTIVE")
	err = retry.Do(
		func() error {
			in, err := p.powervsService.instanceClient.Get(*ins.PvmInstanceID)
			if err != nil {
				return fmt.Errorf("failed to get the instance: %v", err)
			}

			if *in.Status == "ERROR" {
				return fmt.Errorf("instance is in error state")
			}

			if *in.Status == "ACTIVE" {
				logger.Printf("instance is in desired state: %s", *in.Status)
				return nil
			}

			return fmt.Errorf("failed to reach desired instance state")
		},
		retry.Context(ctx),
		retry.Attempts(0),
	)
```
I'm seeing a delay of around 60s before the function returns even though the instance enters the `ACTIVE` state. Not sure if there's an internal delay causing it, I've tried a couple of retry options with no luck. Any idea?

Reply:
@Amulyam24 did you try `retry.MaxDelay(time.Second)`?

`BackoffDelay` is used by default, so delay time increases exponentially like 0.1s, 0.2s, 0.4s, 0.8s, 1.6s, 3.2s, 6.4s, 12.8s, 25.6s, 51.2s, ... `MaxDelay` sets an upper limit for delay time.

Reply:
That worked, Thanks @yoheiueda!